source/v5/mqtt5_to_mqtt3_adapter.c

```diff
@@ -2228,50 +2233,51 @@ static void s_aws_mqtt5_to_mqtt3_adapter_subscribe_completion_fn(
-
-            struct aws_array_list multi_sub_list;
-            aws_array_list_init_static(
-                &multi_sub_list,
-                multi_sub_subscription_ptr_buf,
-                suback->reason_code_count,
-                sizeof(struct aws_mqtt_topic_subscription *));
-
-            size_t subscription_count = aws_array_list_length(&subscribe_op->subscriptions);
-
-            for (size_t i = 0; i < suback->reason_code_count; ++i) {
-                struct aws_mqtt_topic_subscription *subscription = subscription_ptr + i;
-                AWS_ZERO_STRUCT(*subscription);
-
-                subscription->qos = s_convert_mqtt5_suback_reason_code_to_mqtt3_granted_qos(suback->reason_codes[i]);
-
-                if (i < subscription_count) {
-                    aws_array_list_get_at(&subscribe_op->subscriptions, &record, i);
+        // If the suback does not contains any data, we directly extract the data from subscribe_op
+        if (reason_code_count == 0) {
+            reason_code_count = subscription_count;
+        }
 
-                    subscription->topic = record->subscription_view.topic_filter;
-                    subscription->on_publish = record->subscription_view.on_publish_received;
-                    subscription->on_publish_ud = record->subscription_view.callback_user_data;
-                    subscription->on_cleanup = record->subscription_view.on_cleanup;
-                }
+        AWS_VARIABLE_LENGTH_ARRAY(struct aws_mqtt_topic_subscription, multi_sub_subscription_buf, reason_code_count);
+        AWS_VARIABLE_LENGTH_ARRAY(
+            struct aws_mqtt_topic_subscription *, multi_sub_subscription_ptr_buf, reason_code_count);
+        struct aws_mqtt_topic_subscription *subscription_ptr =
+            (struct aws_mqtt_topic_subscription *)multi_sub_subscription_buf;
+
+        struct aws_array_list multi_sub_list;
+        aws_array_list_init_static(
+            &multi_sub_list,
+            multi_sub_subscription_ptr_buf,
+            reason_code_count,
+            sizeof(struct aws_mqtt_topic_subscription *));
+
+        for (size_t i = 0; i < reason_code_count; ++i) {
+            struct aws_mqtt_topic_subscription *subscription = subscription_ptr + i;
+            AWS_ZERO_STRUCT(*subscription);
+
+            if (i < subscription_count) {
+                aws_array_list_get_at(&subscribe_op->subscriptions, &record, i);
+
+                subscription->topic = record->subscription_view.topic_filter;
+                subscription->on_publish = record->subscription_view.on_publish_received;
+                subscription->on_publish_ud = record->subscription_view.callback_user_data;
+                subscription->on_cleanup = record->subscription_view.on_cleanup;
+            }
 
-                aws_array_list_push_back(&multi_sub_list, &subscription);
+            if (suback != NULL) {
+                subscription->qos = s_convert_mqtt5_suback_reason_code_to_mqtt3_granted_qos(suback->reason_codes[i]);
```

Comment:
Debatable: If `suback` is not null then we received it from the server, and according to the mqtt standard the size of `suback->reason_codes` must be equal to the number of subscriptions. However, on the line 2258 we handle the case `length(suback->reason_codes) > subscription_count`. For consistency, we should handle here the case `suback != null && length(suback->reason_codes) == 0`.
Or we can verify that `length(suback->reason_codes) == subscription_count` prior to this loop. And if it's not true, idk, at least add error log.

Reply:
Debatable: I think instead of setting `reason_code_count = subscription_count` in the event of `reason_code_count == 0`, we should just use `subscription_count` in the creation of arrays and iterating in this loop. Then we could use `if (i < reason_code_count)` here which would account for both a null suback and `length(suback->reason_codes) ==0` as well as being able to remove the `if (i < subscription_count)` check on 2258 since the loop is already doing that check.

Reply:
> Debatable: If `suback != null && length(suback->reason_codes) == 0`, we'll get out of bounds access here.
> 
> It's debatable becuase when `suback` is not null then we received it from the server, and according to the mqtt standard the size of `suback->reason_codes` must be equal to the number of subscriptions. However, on the line 2258 we handle the case `length(suback->reason_codes) > subscription_count`, i.e. we handle the situation when the received `suback` packet violates the standard. So, for consistency, we should handle here the case `suback != null && length(suback->reason_codes) == 0`. Or, maybe it'll be better to verify that `length(suback->reason_codes) == subscription_count` prior to this loop. And if it's not true, idk, at least add error log.

We don't need to worry about out of bounds here because the loop will loop through the `reason_code_count` as long as `suback!=NULL`. If `suback->reason_codes) == 0`, we will not go into the loop here.